File: lt1590/Lt1590.java
import java.util.HashMap;   

public class Lt1590{

    public static void main(String args[]){
        Solution sol = new Solution();

        System.out.println(sol.minSubarray( new int[]{ 3,1,4,2 }, 6 )); // Output: 1
        System.out.println(sol.minSubarray( new int[]{6,3,5,2 }, 9 )); // Output: 2
        System.out.println(sol.minSubarray( new int[]{ 1,2,3 }, 3 )); // Output: 0

    }
}

class Solution {
    public int minSubarray(int[] nums, int p) {
        long total = 0;
        for (int x : nums) total += x;

        int rem = (int)(total % p);
        if (rem == 0) return 0; // already divisible

        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);  // prefix before starting

        long prefix = 0;
        int ans = nums.length;

        for (int j = 0; j < nums.length; j++) {
            prefix = (prefix + nums[j]) % p;

            int needed = (int)((prefix - rem + p) % p);

            if (map.containsKey(needed)) {
                ans = Math.min(ans, j - map.get(needed));
            }

            map.put((int)prefix, j);
        }

        return ans == nums.length ? -1 : ans;
    }
}

File: lt1590/lt-1590.js
var minSubarray = function(nums, p) {
    
    const sum = nums.reduce((a,c)=>a+c,0);

    var rem = sum%p ;
    if( rem == 0 ) return 0;
    if( sum < p ) return -1;

    const map = new Map();
    map.set(0,-1);
    var pre = 0;
    var ans = nums.length;

    for(var i=0;i<nums.length;i++)
    {
        pre = (pre+nums[i])%p;
        map.set(pre,i);
        var needed = (pre-rem+p)%p;
        if( map.has(needed) )
            ans = Math.min( ans, i-map.get(needed) );
    }

    return ans==nums.length ? -1 : ans;
};


console.log(minSubarray( [3,1,4,2], 6 )); // Output: 1
console.log(minSubarray( [6,3,5,2], 9 )); // Output: 2
console.log(minSubarray( [1,2,3], 3 )); // Output: 0